### PR TITLE
Bug 9: Update VariadicBase and related to have thread safe

### PR DESCRIPTION
**Pull request by nguhoa**
Fix bug 9 in issue: https://github.com/microsoft/Power-Fx/issues/1535.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Syntax/Nodes/VariadicBase.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public abstract class VariadicBase : TexlNode
         /// </summary>
         public IReadOnlyList<TexlNode> ChildNodes => Children;
 
-        internal readonly TexlNode[] Children;
+        internal readonly IReadOnlyList<TexlNode> Children;
```
Comment (nguhoa) [inline src/libraries/Microsoft.PowerFx.Core/Syntax/Nodes/VariadicBase.cs:23]: >internal readonly IReadOnlyList<TexlNode> Children; [](http://example.com/codeflow?start=8&length=51)

This is key change in this PR. Other places are dependency updated.

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs [hunk 2]
```diff
@@ -4006,7 +4006,7 @@ private ScopeUseSet GetCallNodeScopeUseSet(CallNode node, CallInfo info)
                 }
                 else
                 {
-                    var args = node.Args.Children;
+                    var args = node.Args.Children.ToArray();
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4009]: >.ToArray() [](http://example.com/codeflow?start=49&length=10)

no need to `ToArray` here - can use .Count on line 4012

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs [hunk 3]
```diff
@@ -4328,7 +4328,7 @@ public override bool PreVisit(CallNode node)
                 // The zeroth arg should not be a lambda. Instead it defines the context type for the lambdas.
                 Contracts.Assert(!maybeFunc.IsLambdaParam(0));
 
-                var args = node.Args.Children;
+                var args = node.Args.Children.ToArray();
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4331]: >.ToArray() [](http://example.com/codeflow?start=45&length=10)

No need to `ToArray` - can use args.Count where args.Length is used.

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs [hunk 4]
```diff
@@ -4801,7 +4801,7 @@ private void PreVisitMetadataArg(CallNode node, TexlFunction func)
                 Contracts.Assert(func.SupportsMetadataTypeArg);
                 Contracts.Assert(!func.HasLambdas);
 
-                var args = node.Args.Children;
+                var args = node.Args.Children.ToArray();
```
[the remark this answers is not in the document]
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4804]: HandleCheckInvocation in 4855 need to pass in arrary. so either we have to toarray here or toarrary in line 4855.

[the remark this answers is not in the document]
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4804]: Ah, you're right, missed that one.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**